service/storage_service.cc

```diff
@@ -5540,25 +5534,42 @@ future<> storage_service::stream_tablet(locator::global_tablet_id tablet) {
-        auto streamer = make_lw_shared<dht::range_streamer>(_db, _stream_manager, std::move(tm),
-                                                            guard.get_abort_source(),
-                                                            tm->get_my_id(), _snitch.local()->get_location(),
-                                                            format("Tablet {}", trinfo->transition),
-                                                            reason,
-                                                            topo_guard,
-                                                            std::move(tables));
-        tm = nullptr;
-        streamer->add_source_filter(std::make_unique<dht::range_streamer::failure_detector_source_filter>(
-                _gossiper.get_unreachable_members()));
-
-        std::unordered_map<inet_address, dht::token_range_vector> ranges_per_endpoint;
-        for (auto r : streaming_info.read_from) {
-            ranges_per_endpoint[host2ip(r.host)].emplace_back(range);
+        if (trinfo->transition == locator::tablet_transition_kind::intranode_migration) {
+            locator::tablet_replica leaving_replica = locator::get_leaving_replica(tinfo, *trinfo);
+            auto range = to_partition_range(tmap.get_token_range(tablet.tablet));
+            co_await smp::submit_to(leaving_replica.shard, [this, tablet, pending_replica, reason, topo_guard, range] () -> future<> {
+                rtlogger.info("Intra-node tablet streaming of {} to {}", tablet, pending_replica.shard);
+                auto& table = _db.local().find_column_family(tablet.table);
+                dht::auto_refreshing_sharder sharder(table.shared_from_this());
+                auto estimated_partitions = 1; // FIXME
+                auto permit = co_await _db.local().obtain_reader_permit(table, "stream-session", db::no_timeout, {});
+                co_await mutation_writer::distribute_reader_and_consume_on_shards(table.schema(), sharder,
+                    table.make_streaming_reader(table.schema(), std::move(permit), range, gc_clock::now()),
+                    _stream_manager.local().make_streaming_consumer(estimated_partitions, reason, topo_guard),
+                    table.stream_in_progress());
+                rtlogger.info("Intra-node tablet streaming of {} to {} finished", tablet, pending_replica.shard);
+            });
+        } else {
+            auto& table = _db.local().find_column_family(tablet.table);
+            std::vector<sstring> tables = {table.schema()->cf_name()};
+            auto streamer = make_lw_shared<dht::range_streamer>(_db, _stream_manager, std::move(tm),
+                                                                guard.get_abort_source(),
+                                                                tm->get_my_id(), _snitch.local()->get_location(),
+                                                                format("Tablet {}", trinfo->transition),
```

Comment:
undefined behavior in code above, but unrelated to your series, will send a fix.